scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+            elif typecode == cnp.NPY_CFLOAT:
+                blas_t_2d_conj(m, p, <float_complex*>uvoid, us)
+                gemm(N, utrans, m, p, m, 1, <float_complex*>qvoid, qs[0],
+                        <float_complex*>uvoid, ldu, 0,
+                        col(<float_complex*>qTuvoid, qTus, k), qTus[1])
+                blas_t_2d_conj(m, p, col(<float_complex*>qTuvoid, qTus, k), qTus)
+            elif typecode == cnp.NPY_CDOUBLE:
+                blas_t_2d_conj(m, p, <double_complex*>uvoid, us)
+                gemm(N, utrans, m, p, m, 1, <double_complex*>qvoid, qs[0],
+                        <double_complex*>uvoid, ldu, 0,
+                        col(<double_complex*>qTuvoid, qTus, k), qTus[1])
+                blas_t_2d_conj(m, p, col(<double_complex*>qTuvoid, qTus, k), qTus)
+        else:
+            raise ValueError('1 <= u.ndim <= 2')
+    else:
+        raise ValueError('q must be either F or C contig')
+
+cdef validate_array(cnp.ndarray a, bint chkfinite):
+    # here we check that a has positive strides and that its size is small
+    # enough to fit in into an int, as BLAS/LAPACK require
+    cdef bint copy = False
+    cdef bint too_large = False
+    cdef int j
+
+    for j in range(a.ndim):
+        if a.strides[j] <= 0 or \
+                (a.strides[j] / a.descr.itemsize) >= limits.INT_MAX:
+            copy = True
+        if a.shape[j] >= limits.INT_MAX:
+            raise ValueError('Input array to large for use with BLAS')
+
+    if chkfinite:
+        if not np.isfinite(a).all():
+            raise ValueError('array must not contain infs or NaNs')
+
+    if copy:
+            return PyArray_FromArraySafe(a, NULL, cnp.NPY_F_CONTIGUOUS)
+    return a
+
+cdef validate_qr(object q0, object r0, bint overwrite_q, int q_order,
+                 bint overwrite_r, int r_order, bint chkfinite):
+    cdef cnp.ndarray Q
+    cdef cnp.ndarray R
+    cdef int typecode
+    cdef bint economic = False
+
+    q_order |= cnp.NPY_BEHAVED_NS | cnp.NPY_ELEMENTSTRIDES  
+    r_order |= cnp.NPY_BEHAVED_NS | cnp.NPY_ELEMENTSTRIDES
+
+    if not overwrite_q:
+        q_order |= cnp.NPY_ENSURECOPY
+
+    if not overwrite_r:
+        r_order |= cnp.NPY_ENSURECOPY
+
+    # in the interests of giving better error messages take any number of
+    # dimensions here.
+    Q = PyArray_CheckFromAny(q0, NULL, 0, 0, q_order, NULL)
+    R = PyArray_CheckFromAny(r0, NULL, 0, 0, r_order, NULL)
+
+    if Q.ndim != 2 or R.ndim != 2:
+        raise ValueError('Q and R must be 2d')
+
+    typecode = cnp.PyArray_TYPE(Q)
+
+    if typecode != cnp.PyArray_TYPE(R):
+        raise ValueError('q and r must have the same type')
+
+    if not (typecode == cnp.NPY_FLOAT or typecode == cnp.NPY_DOUBLE 
+            or typecode == cnp.NPY_CFLOAT or typecode == cnp.NPY_CDOUBLE):
+        raise ValueError('only floatingcomplex arrays supported')
+
+    # we support MxM MxN and MxN NxN
```

Comment:
would be good to say this in both error messages below: ``"expected (M,M) (M,N) or (M,N) (N,N) but found (...)"``